.circleci/config.yml

```
@@ -93,6 +93,75 @@ commands:
             - bytecode-report-<< parameters.label >>-cli.txt
       - matrix_notify_failure_unless_pr
 
+  install_python3:
+    description: "Install python3 and given packages."
+    parameters:
+      packages:
+        description: "List of optional python packages to be installed (separated by space)."
+        type: string
+        default: ""
+    steps:
+      - run:
+          name: Install python and dependencies
+          command: |
+            sudo apt update
+            if ! python3 --version >/dev/null 2>&1
+            then
+              sudo apt install python3 --assume-yes --no-install-recommends
+            fi
+            sudo apt install python3-pip
+            if [ -n "<< parameters.packages >>" ]
+            then
+              echo "Installing additional packages..."
+              python3 -m pip install --user << parameters.packages >>
+            fi
+
+  install_foundry:
+    description: "Install Foundry."
+    parameters:
+      version:
+        type: string
+        default: "nightly"
+      install_path:
+        type: string
+        default: /home/circleci/bin
+    steps:
+      - run:
+          name: Setup Foundry environment variables
+          command: |
+            FOUNDRY_REPO="foundry-rs/foundry"
+            FOUNDRY_VERSION=<< parameters.version >>
+            FOUNDRY_RELEASE_SHA=$(curl \
+              --silent \
+              --fail \
+              --show-error \
+              "https://api.github.com/repos/${FOUNDRY_REPO}/git/refs/tags/${FOUNDRY_VERSION}" \
+              | jq --raw-output .object.sha \
+            )
+            echo "export FOUNDRY_REPO=$FOUNDRY_REPO" >> "$BASH_ENV"
+            echo "export FOUNDRY_VERSION=$FOUNDRY_VERSION" >> "$BASH_ENV"
+            echo "export FOUNDRY_RELEASE_TAG='nightly-${FOUNDRY_RELEASE_SHA}'" >> "$BASH_ENV"
+            # Save commit sha for caching
+            echo $FOUNDRY_RELEASE_SHA > /tmp/workspace/foundry-release-sha
+      - restore_cache:
+          keys:
+            - foundry-<< parameters.version >>-{{ checksum "/tmp/workspace/foundry-release-sha" }}
+      - run:
+          name: Install foundry
+          command: |
+            if ! command -v forge > /dev/null 2>&1; then
+              wget \
+                --output-document /tmp/foundry.tar.gz \
+                "https://github.com/${FOUNDRY_REPO}/releases/download/${FOUNDRY_RELEASE_TAG}/foundry_${FOUNDRY_VERSION}_linux_amd64.tar.gz"
```

It's not like it wrong, but it's kinda weird to randomly switch between `curl` and `wget`. Why not stick to one?

```suggestion
              curl \
                --output /tmp/foundry.tar.gz \
                "https://github.com/${FOUNDRY_REPO}/releases/download/${FOUNDRY_RELEASE_TAG}/foundry_${FOUNDRY_VERSION}_linux_amd64.tar.gz"
```

---

No big reason, I usually use `wget` to download files from the web for better error reporting and useful defaults. For instance, in the way you suggested, `curl` would download an empty file (because it does not follow redirections by default) or in case of errors, it would download a text file containing an error string, while `wget` would immediately report `404`. Of course this can be fixed by adding other parameters to `curl` like `--fail --location` but yeah.